### PR TITLE
p-866 Adding btc token holding amount

### PR DESCRIPTION
**Pull request by 0xverin**
### Context

as the topic;
this pr also adding dataprovider client to tokeninfo structure

**The CLI test has not passed yet. But it can be reviewed first.**

### Labels

Please apply following PR-related labels when appropriate:
- `C0-breaking`: if your change could break the existing client, e.g. API change, critical logic change 
- `C1-noteworthy`: if your change is non-breaking, but is still worth noticing for the client, e.g. reference code improvement

### Testing Evidences
```
./bin/litentry-cli trusted -d link-identity did:litentry:substrate:0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d did:litentry:bitcoin:0x0381a0d6b97a0e65d1268b3ea2cd47d7cf3387b2811116b88da88cba11c6584c9e bitcoinp2wpkh
```

```
./bin/litentry-cli trusted -d request-vc did:litentry:substrate:0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d -a "dynamic 0000000000000000000000000000000000000000 000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000036274630000000000000000000000000000000000000000000000000000000000 true"
>>> identity: Substrate(Address32)
>>> nonce: 1
>>> assertions: [Dynamic(DynamicParams { smart_contract_id: 0x0000000000000000000000000000000000000000, smart_contract_params: Some(BoundedVec([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 32, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 98, 116, 99, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], 1024)), return_log: true })]
```
```
{
    "@context": [
        "https://www.w3.org/2018/credentials/v1",
        "https://w3id.org/security/suites/ed25519-2020/v1"
    ],
    "id": "0x181cd1661794d3eb5ccbca020dfa58911ebae5c2a79d979bc0db4f85879f89ac",
    "type": [
        "VerifiableCredential"
    ],
    "credentialSubject": {
        "id": "did:litentry:substrate:0xd43593c715fdd31c61141abd04a99fd6822c8558854ccde39a5684e7a56da27d",
        "description": "The amount of a particular token you are holding",
        "type": "Token Holding Amount",
        "assertionText": "",
        "assertions": [
            {
                "and": [
                    {
                        "src": "$token",
                        "op": "==",
                        "dst": "btc"
                    },
                    {
                        "src": "$holding_amount",
                        "op": ">=",
                        "dst": "5"
                    },
                    {
                        "src": "$holding_amount",
                        "op": "<",
                        "dst": "10"
                    }
                ]
            }
        ],
        "values": [
            true
        ],
        "endpoint": "http://localhost:9933/"
    },
    "issuer": {
        "id": "did:litentry:substrate:0xe4f0691ecba8153794085269c53259a3250222b6edd42ce4d792f39f9f2db6c3",
        "name": "Litentry TEE Worker",
        "mrenclave": "6T6AVMxXs9SbQh8Zp9yCgDbRWSUGYYHFeY4xhm63B8Av",
        "runtimeVersion": {
            "parachain": 9186,
            "sidechain": 109
        }
    },
    "issuanceDate": "2024-07-31T08:06:28.431767421+00:00",
    "parachainBlockNumber": 474,
    "sidechainBlockNumber": 943,
    "proof": {
        "created": "2024-07-31T08:06:28.432700166+00:00",
        "type": "Ed25519Signature2020",
        "proofPurpose": "assertionMethod",
        "proofValue": "cf2ab0cd98cfa717f30432b17cd157deb07a59c9c4c674230f70027b7fc086efed9b7e59338ec246e22d7429b01725bdbca5a09afbe7e0700152d1f717250506",
        "verificationMethod": "0x17426351f9a0d1718c3746d84c60e5cd99352312914663a679c14bbea10f7812"
    },
    "credentialSchema": {
        "id": "https://raw.githubusercontent.com/litentry/vc-jsonschema/main/dist/schemas/25-token-holding-amount/1-1-3.json",
        "type": "JsonSchemaValidator2018"
    }
}
```

image

### CODE REVIEW

**Review by kziemianek — APPROVED** [review]

**Review by kziemianek — COMMENTED** [review]
Could you add hardhat tests ?

**Comment by 0xverin** [comment]
> add hardhat tests ?

yes

**Comment by 0xverin** [comment]
> Could you add hardhat tests ?

Okay, that's why I didn't merge after approve.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Dai.sol [hunk 1]
```diff
@@ -41,15 +41,18 @@ library Dai {
...
 			Web3Networks.Bsc,
-			"0x1af3f329e8be154074d8769d1ffa4ee058b1dbc3"
+			"0x1af3f329e8be154074d8769d1ffa4ee058b1dbc3",
+			DataProviderTypes.NoderealClient
 		);
 		tokenInfoList[2] = TokenInfo(
 			Web3Networks.Solana,
-			"EjmyN6qEC1Tf1JxiG1ae7UTJhUxSwk1TCWNWqxWV4J6o"
+			"EjmyN6qEC1Tf1JxiG1ae7UTJhUxSwk1TCWNWqxWV4J6o",
+			DataProviderTypes.NoderealClient
```
Comment (wli-pro) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Dai.sol]: Nodereal may not support Solana network, pls check and see if other existing data provider support solana.

**Review by wli-pro — CHANGES_REQUESTED** [review]

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Dai.sol (hunk 1 above)
Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Dai.sol]: Done. All changes to Moralis.

File: tee-worker/litentry/core/assertion-build/src/dynamic/v20.14.0
Comment (wli-pro) [inline tee-worker/litentry/core/assertion-build/src/dynamic/v20.14.0:1]: Any reason to keep this in git or we should remove it?

Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/v20.14.0:1]: Removed.

File: tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Cro.sol [hunk 2]
```diff
@@ -36,15 +36,19 @@ library Cro {
...
+            "0xa0b73e1ff0b80914ab6fe0444e65848c4c34450b",
+            DataProviderTypes.NoderealClient,
+            18
         );
-        networks[1] = TokenInfo(
+        tokenInfoList[1] = TokenInfo(
             Web3Networks.Solana,
-            "DvjMYMVeXgKxaixGKpzQThLoG98nc7HSU7eanzsdCboA"
+            "DvjMYMVeXgKxaixGKpzQThLoG98nc7HSU7eanzsdCboA",
+            DataProviderTypes.NoderealClient,
```
Comment (wli-pro) [inline tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Cro.sol]: Nodereal shouldn't support Solana network?

Comment (0xverin) [reply tee-worker/litentry/core/assertion-build/src/dynamic/contracts/token_holding_amount/erc20/Cro.sol]: my bad , one was missed.

**Comment by wli-pro** [comment]
LGTM, could you pls add some cli test evidence?

**Comment by 0xverin** [comment]
> LGTM, could you pls add some cli test evidence?

pls see description.

**Comment by wli-pro** [comment]
> > LGTM, could you pls add some cli test evidence?
> 
> pls see description.

Does the true case also work?

**Comment by 0xverin** [comment]
> > > LGTM, could you pls add some cli test evidence?
> > 
> > 
> > pls see description.
> 
> Does the true case also work?

Done!

> > > LGTM, could you pls add some cli test evidence?
> > 
> > 
> > pls see description.
> 
> Does the true case also work?

Updated.

**Review by wli-pro — APPROVED** [review]

**Pull request merged**